### PR TITLE
Regularise era based command structure

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    Regularise era based command structure
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

The main change in this PR is to change command parsers to return `Maybe (Parser a)` instead of `Parser a`.

This is based on the principle that commands themselves should determined whether they exist in a given era.  This principle is used in the new Conway commands and will see more use in the future.

Requiring that all command groups accept `Maybe (Parser a)` will mean that any new or existing commands can become era-sensitive without inducing refactoring.  As it is at the moment if a command needs to become era-sensitive, the parent command group will need to be refactored to accomodate it if it doesn't already which is the case for a number top-level commands groups.

The PR also contains a small number of help text fixes.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - round trip tests
  - integration tests
  See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

### CODE REVIEW

**Review by carbolymer — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs [hunk 1]
```diff
@@ -14,22 +14,27 @@ import           Cardano.CLI.EraBased.Commands.Address
...
 import           Data.Foldable
+import           Data.Maybe
 import           Options.Applicative hiding (help, str)
 import qualified Options.Applicative as Opt
 
 pAddressCmds :: CardanoEra era -> EnvCli -> Parser (AddressCmds era)
 pAddressCmds _ envCli =
-  asum
-    [ subParser "key-gen"
+  asum $ catMaybes
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs]: I don't think this is necessary as these commands will be available in every era unless e.g the crypto changes.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs [hunk 2]
```diff
@@ -27,50 +27,61 @@ import qualified Options.Applicative as Opt
 
 pGenesisCmds :: EnvCli -> Parser (GenesisCmds era)
 pGenesisCmds envCli =
-  asum
-    [ subParser "key-gen-genesis"
+  asum $ catMaybes
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs]: I'm not sure I understand the reason for this change.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs [hunk 3]
```diff
@@ -13,23 +13,34 @@ import           Cardano.CLI.Environment (EnvCli (..))
...
-import           Data.Foldable
 import           Options.Applicative hiding (help, str)
 import qualified Options.Applicative as Opt
 
-pAddressCmds :: CardanoEra era -> EnvCli -> Parser (AddressCmds era)
+pAddressCmds :: ()
+  => CardanoEra era
+  -> EnvCli
+  -> Maybe (Parser (AddressCmds era))
 pAddressCmds _ envCli =
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs:23]: So we're not using the `CardanoEra` value here because these commands are constant for all the eras. I'm assuming you're doing this in the event that e.g the cryptography changes in a future era?

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs [hunk 4]
```diff
@@ -16,61 +16,78 @@ import           Cardano.CLI.EraBased.Options.Common
...
 pGenesisCmds envCli =
-  asum
-    [ subParser "key-gen-genesis"
+  subInfoParser "genesis"
+    ( Opt.progDesc
+        $ mconcat
+          [ "Genesis block commands."
+          ]
+    )
+    [ Just
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs:37]: Why the change from `asum`?

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Query.hs [hunk 5]
```diff
@@ -14,37 +14,36 @@ import           Cardano.CLI.EraBased.Options.Common
...
 
-pGovernanceQueryGetConstitutionCmd
-  :: EnvCli
-  -> CardanoEra era
+pGovernanceQueryGetConstitutionCmd :: ()
+  => CardanoEra era
+  -> EnvCli
   -> Maybe (Parser (GovernanceQueryCmds era))
-pGovernanceQueryGetConstitutionCmd env era = do
+pGovernanceQueryGetConstitutionCmd era env = do
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Query.hs:35]: :+1:

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Key.hs [hunk 6]
```diff
@@ -21,65 +21,78 @@ import qualified Options.Applicative as Opt
 {- HLINT ignore "Use <$>" -}
 {- HLINT ignore "Move brackets to avoid $" -}
 
-pKeyCmds :: Parser (KeyCmds era)
+pKeyCmds :: Maybe (Parser (KeyCmds era))
 pKeyCmds =
-  asum
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Key.hs:26]: Why the change from `Parser (..)` -> `Maybe (Parser (...))`

**Review by Jimbo4350 — DISMISSED** [review]
Some questions

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs (hunk 3 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs:23]: Or if we add new commands that are only apply to a future era.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Key.hs (hunk 6 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Key.hs:26]: The change represents a shift of responsibilities.

Previously it was the parent command group that decides whether the command should appear in an era.

The change shifts the responsibility from the parent to the command in question.  ie the command itself determines whether or not it should show up in a given era.

This is already how it is done for the governance commands and some other commands.

The purpose of this PR is to harmonise our code base to consistently use the latter convention.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs (hunk 3 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Address.hs:23]: Some scenarios that might cause the era to be need:

* A new command or flag added that only apply to newer eras
* An existing command or flag needs to be deleted for newer eras

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs (hunk 4 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs:37]: `asum` is in the implementation of `subInfoParser` so we are still calling `asum` indirectly:

```
subInfoParser :: String -> InfoMod a -> [Maybe (Parser a)] -> Maybe (Parser a)
subInfoParser name i mps = case catMaybes mps of
  [] -> Nothing
  parsers -> Just $ subParser name $ Opt.info (asum parsers) i
```

`subInfoParser` is used instead so that commands are self-describing.  Previously the implementation of the parent command group provided the description.
